titiler/cmr/timeseries.py

```diff
@@ -0,0 +1,612 @@
+from fastapi import Body, Depends, Path, Query, Request, Response
+from fastapi.exceptions import HTTPException
+from fastapi.responses import StreamingResponse
+from geojson_pydantic import Feature, FeatureCollection
+from geojson_pydantic.geometries import Geometry
+from PIL import Image
+from pydantic import BaseModel
+
+from titiler.cmr.dependencies import ConceptID
+from titiler.cmr.factory import Endpoints
+from titiler.core.algorithm import algorithms as available_algorithms
+from titiler.core.dependencies import CoordCRSParams, DefaultDependency, DstCRSParams
+from titiler.core.factory import FactoryExtension
+from titiler.core.models.mapbox import TileJSON
+from titiler.core.models.responses import Statistics
+from titiler.core.resources.enums import ImageType
+from titiler.core.resources.responses import GeoJSONResponse
+
+timeseries_img_endpoint_params: Dict[str, Any] = {
+    "responses": {
+        200: {
+            "content": {
+                "image/gif": {},
+            },
+            "description": "Return an image.",
+        }
+    },
+    "response_class": Response,
+}
+
+
+class TimeseriesMediaType(str, Enum):
+    """Responses Media types formerly known as MIME types."""
+
+    gif = "image/gif"
+
+
+class TimeseriesImageType(str, Enum):
+    """Available output image type."""
+
+    gif = "gif"
+
+    @DynamicClassAttribute
+    def mediatype(self):
+        """Return image media type."""
+        return TimeseriesMediaType[self._name_].value
+
+
+TimeseriesStatistics = Dict[str, Statistics]
+
+
+class TimeseriesStatisticsInGeoJSON(BaseModel):
+    """Statistics model in geojson response."""
+
+    statistics: TimeseriesStatistics
+
+    model_config = {"extra": "allow"}
+
+
+TimeseriesStatisticsGeoJSON = Union[
+    FeatureCollection[Feature[Geometry, TimeseriesStatisticsInGeoJSON]],
+    Feature[Geometry, TimeseriesStatisticsInGeoJSON],
+]
+
+TimeseriesTileJSON = Dict[str, TileJSON]
+
+
+@dataclass
+class TimeseriesParams(DefaultDependency):
+    """Timeseries parameters"""
+
+    start_datetime: Annotated[
+        Optional[str],
+        Query(
+            description="Start datetime for timeseries request",
+        ),
+    ] = None
+    end_datetime: Annotated[
+        Optional[str],
+        Query(
+            description="End datetime for timeseries request",
+        ),
+    ] = None
+    step: Annotated[
+        Optional[str],
+        Query(
+            description="Time step between timeseries intervals, expressed as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations)"
+        ),
+    ] = None
+    step_idx: Annotated[
+        Optional[int],
+        Query(description="Optional (zero-indexed) index of the desired time step"),
+    ] = None
+    exact: Annotated[
+        Optional[bool],
+        Query(
+            description="If true, queries will be made for a point-in-time at each step. If false, queries will be made for the entire interval between steps"
```

Comment:
Yes that is a great point. I was thinking about adding some kind of explainer in the notebook but you are right - at no point do we perform an average over multiple assets in a time range. The [`mosaic_reader`](https://github.com/cogeotiff/rio-tiler/blob/9a39c802044b485a26c29065b539ec44c817d8a6/rio_tiler/mosaic/reader.py#L25) can take a list of assets but they get flattened to a single 2D array (per variable) and I am 90% sure it will just fill in the array one asset at a time, and if they overlap the final values come from the last asset.

I will add some more annotations to the arguments here and some details in the notebook to explain that better.

Reply:
>  and if they overlap the final values come from the last asset.

From the first asset! (the default mosaic selection method is `FirstMethod` https://github.com/cogeotiff/rio-tiler/blob/9a39c802044b485a26c29065b539ec44c817d8a6/rio_tiler/mosaic/methods/defaults.py#L10-L27